src/typecheck.rs

```diff
@@ -78,368 +294,166 @@ impl<T> Atom<T> {
     }
 
-    fn accept(&self, types: &[ArcSort]) -> Option<ArcSort> {
-        match types {
-            [a, b] if a.name() == b.name() => Some(a.clone()),
-            _ => None,
-        }
+    fn get_type_constraints(&self) -> Box<dyn TypeConstraint> {
+        AllEqualTypeConstraint::new(self.name())
+            .with_exact_length(3)
+            .with_output_sort(self.unit.clone())
+            .into_box()
     }
 
     fn apply(&self, values: &[Value]) -> Option<Value> {
         assert_eq!(values.len(), 2);
         if values[0] == values[1] {
-            Some(values[0])
+            Some(Value::unit())
         } else {
             None
         }
     }
 }
 
 impl<'a> Context<'a> {
-    pub fn new(egraph: &'a EGraph) -> Self {
-        Self {
-            egraph,
-            unit: egraph.type_info().sorts[&Symbol::from(UNIT_SYM)].clone(),
-            types: Default::default(),
-            errors: Vec::default(),
-            unionfind: UnionFind::default(),
-            nodes: HashMap::default(),
-        }
+    pub fn new(egraph: &'a mut EGraph) -> Self {
+        Self { egraph }
     }
 
-    fn add_node(&mut self, node: ENode) -> Id {
-        let entry = self.nodes.entry(node);
-        *entry.or_insert_with(|| self.unionfind.make_set())
-    }
-
-    pub fn typecheck_query(
-        &mut self,
-        facts: &'a [Fact],
-        actions: &'a [Action],
-    ) -> Result<(Query, Vec<Action>), Vec<TypeError>> {
-        for fact in facts {
-            self.typecheck_fact(fact);
-        }
-
-        // congruence isn't strictly necessary, but it can eliminate some redundant atoms
-        self.rebuild();
-
-        // First find the canoncial version of each leaf
-        let mut leaves = HashMap::<Id, Expr>::default();
-        let mut canon = HashMap::<Symbol, Expr>::default();
-
-        // Do literals first
-        for (node, &id) in &self.nodes {
-            match node {
-                ENode::Literal(lit) => {
-                    let old = leaves.insert(id, Expr::Lit(lit.clone()));
-                    if let Some(Expr::Lit(old_lit)) = old {
-                        panic!("Duplicate literal: {:?} {:?}", old_lit, lit);
-                    }
-                }
-                _ => continue,
-            }
-        }
-        // Globally bound variables first
-        for (node, &id) in &self.nodes {
-            match node {
-                ENode::Var(var) => {
-                    if self.egraph.global_bindings.get(var).is_some() {
-                        match leaves.entry(id) {
-                            Entry::Occupied(existing) => {
-                                canon.insert(*var, existing.get().clone());
-                            }
-                            Entry::Vacant(v) => {
-                                v.insert(Expr::Var(*var));
-                            }
+    // TODO: should this be in ResolvedCoreRule
+    pub fn canonicalize(&self, rule: UnresolvedCoreRule) -> UnresolvedCoreRule {
```

Comment:
Another place where it would be good to add documentation- what's the purpose of canonicalize?

Reply:
Good point. I have documented its purpose.
```

    /// Transformed a UnresolvedCoreRule into a CanonicalizedCoreRule.
    /// In particular, it removes equality checks between variables and
    /// other arguments, and turns equality checks between non-variable arguments
    /// into a primitive equality check `value-eq`.
```

Reply:
FWIW, `value-eq` cannot be completely removed from our codebase since queries may contain equality checks between e.g., global variables, which cannot be compiled away.